stl/inc/atomic

```diff
@@ -1675,6 +1795,101 @@ template <class _Ty>
 atomic(_Ty) -> atomic<_Ty>;
 #endif // _HAS_CXX17
 
+#if _HAS_CXX20
+template <class _Ty>
+struct atomic_ref : _Choose_atomic_base_t<_Ty, _Ty&> { // atomic reference
+private:
+    using _Base = _Choose_atomic_base_t<_Ty, _Ty&>;
+
+public:
+    // clang-format off
+    static_assert(is_trivially_copyable_v<_Ty> && is_copy_constructible_v<_Ty> && is_move_constructible_v<_Ty>
+        && is_copy_assignable_v<_Ty> && is_move_assignable_v<_Ty>,
+        "atomic<T> requires T to be trivially copyable, copy constructible, move constructible, copy assignable, "
+        "and move assignable.");
+    // clang-format on
+
+    using value_type = _Ty;
+
+    explicit atomic_ref(_Ty& _Value) : _Base(_Value) {
+        if constexpr (_Is_potentially_lock_free) {
+            _Check_aligment(_Value);
+        }
+
+#if 1 // TRANSITION, ABI
+        if constexpr (!is_always_lock_free) {
+#else
+        if constexpr (!_Is_potentially_lock_free) {
+#endif
+            this->_Init_spinlock_for_ref();
+        }
+    }
+
+    atomic_ref(const atomic_ref&) noexcept = default;
+
+    atomic_ref& operator=(const atomic_ref&) = delete;
+
+    static constexpr bool is_always_lock_free = _Is_always_lock_free<sizeof(_Ty)>;
+
+    static constexpr bool _Is_potentially_lock_free =
+        sizeof(_Ty) <= 2 * sizeof(void*) && (sizeof(_Ty) & (sizeof(_Ty) - 1)) == 0;
+
+    static constexpr size_t required_alignment = _Is_potentially_lock_free ? sizeof(_Ty) : alignof(_Ty);
+
+#if 1 // TRANSITION, ABI

```

Comment:
There's no existing ABI for `atomic_ref` so we shouldn't need tests like this.

Reply:
There is, since `atomic` machinery is reused. 
If you implement `chpxchg16b` for `atomic_ref` separately from `atomic`, then remove this.

Reply:
If C++ allows further specializing currently #if0ed part as follows:
```
template <class _Ty>
struct _Atomic_storage<_Ty&, 16>
```
Then it should be a good idea

Reply:
I really like the clever "plumb it though `atomic`" thing you've done here but not if it causes us to deliver a suboptimal experience :(

Reply:
I will see, I think it is possible to specialize `atomic_ref<16 bytes>` one way or another.
And it will give a coverage of currently #if0ed case.

Reply:
Maybe expose a bool "is atomic_ref" or similar from `_Atomic_storage_types` which is true for the T& partial specialization, and then `if _CONSTEXPR_IF` against that in `_Atomic_storage<16 bytes>`?

Reply:
```
template <class _Ty>
struct _Atomic_storage<_Ty&, 16> {
```

This works.
Though this
```
#if 0 // TRANSITION, ABI
struct _Atomic_storage<_Ty&, 16> {
#else
struct _Atomic_storage<_Ty&, 16> {
#endif

Confuses intellisense, so I'll put it this way:
```
template <class _Ty>
struct _Atomic_storage<_Ty&, 16> {
    //  TRANSITION, ABI: change _Ty& to _Ty
```

Reply:
Done